metricproviders/wavefront/wavefront.go

```diff
@@ -0,0 +1,240 @@
+)
+
+const (
+	//ProviderType indicates the provider is wavefront
+	ProviderType = "Wavefront"
+	//k8s secret that has wavefront api tokens
+	WavefrontTokensSecretName = "wavefront-api-tokens"
+)
+
+type Provider struct {
+	api    WavefrontClientAPI
+	logCtx log.Entry
+}
+
+func (p *Provider) Type() string {
+	return ProviderType
+}
+
+type WavefrontClientAPI interface {
+	NewQuery(params *wavefront_api.QueryParams) WavefrontQueryAPI
+}
+
+type WavefrontClient struct {
+	*wavefront_api.Client
+}
+
+func (wc *WavefrontClient) NewQuery(params *wavefront_api.QueryParams) WavefrontQueryAPI {
+	return &WavefrontQuery{Query: wc.Client.NewQuery(params)}
+}
+
+type WavefrontQueryAPI interface {
+	Execute() (*wavefront_api.QueryResponse, error)
+}
+
+type WavefrontQuery struct {
+	*wavefront_api.Query
+}
+
+func (wq *WavefrontQuery) Execute() (*wavefront_api.QueryResponse, error) {
+	return wq.Query.Execute()
+}
+
+// Run queries with wavefront provider for the metric
+func (p *Provider) Run(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric) v1alpha1.Measurement {
+	startTime := metav1.Now()
+	newMeasurement := v1alpha1.Measurement{
+		StartedAt: &startTime,
+	}
+
+	query, err := templateutil.ResolveArgs(metric.Provider.Wavefront.Query, run.Spec.Args)
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+
+	queryParams := &wavefront_api.QueryParams{
+		QueryString:             query,
+		StartTime:               strconv.FormatInt(time.Now().Unix()*1000, 10),
+		MaxPoints:               "1",
+		Granularity:             "s",
+		SeriesOutsideTimeWindow: false,
+	}
+
+	response, err := p.api.NewQuery(queryParams).Execute()
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+	newValue, newStatus, err := p.processResponse(metric, response)
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+
+	}
+	newMeasurement.Value = newValue
+	newMeasurement.Phase = newStatus
+	finishedTime := metav1.Now()
+	newMeasurement.FinishedAt = &finishedTime
+	return newMeasurement
+}
+
+// Resume should not be used the Wavefront provider since all the work should occur in the Run method
+func (p *Provider) Resume(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric, measurement v1alpha1.Measurement) v1alpha1.Measurement {
+	p.logCtx.Warn("Wavefront provider should not execute the Resume method")
+	return measurement
+}
+
+// Terminate should not be used the Wavefront provider since all the work should occur in the Run method
+func (p *Provider) Terminate(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric, measurement v1alpha1.Measurement) v1alpha1.Measurement {
+	p.logCtx.Warn("Wavefront provider should not execute the Terminate method")
+	return measurement
+}
+
+// GarbageCollect is a no-op for the Wavefront provider
+func (p *Provider) GarbageCollect(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric, limit int) error {
+	return nil
+}
```

Comment:
Can you add some quick unit tests for these functions and the Type function too? It should be really quick if you use the Prometheus tests as examples for these tests:
Type: https://github.com/argoproj/argo-rollouts/blob/master/metricproviders/prometheus/prometheus_test.go#L27-L34
Resume: https://github.com/argoproj/argo-rollouts/blob/master/metricproviders/prometheus/prometheus_test.go#L129-L150
Terminate: https://github.com/argoproj/argo-rollouts/blob/master/metricproviders/prometheus/prometheus_test.go#L152-L164
GarageCollect: https://github.com/argoproj/argo-rollouts/blob/master/metricproviders/prometheus/prometheus_test.go#L166-L172